src/Spice86.Core/Emulator/Devices/Video/AeonCard.cs

```diff
@@ -0,0 +1,836 @@
+
+    /// <summary>
+    ///     Gets the VGA attribute controller.
+    /// </summary>
+    public AttributeController AttributeController { get; } = new();
+
+    /// <summary>
+    ///     Gets a pointer to the emulated video RAM.
+    /// </summary>
+    public nint VideoRam { get; }
+
+    /// <summary>
+    ///     Gets the text-mode display instance.
+    /// </summary>
+    public TextConsole TextConsole { get; }
+
+    public void SetVramByte(uint address, byte value) {
+        CurrentMode.SetVramByte(address, value);
+    }
+
+    public void Render(uint address, object width, object height, nint pixelsAddress) {
+        _presenter ??= GetPresenter();
+        _presenter.Update(pixelsAddress);
+    }
+
+    public void TickRetrace() {
+        // Inactive at tick time, but will become active once the code checks for it.
+        _crtStatusRegister = StatusRegisterRetraceInactive;
+    }
+
+    public void UpdateScreen() {
+        _gui?.UpdateScreen();
+    }
+
+    public event EventHandler? VideoModeChanged;
+
+    private void FillDispatchTable() {
+        _dispatchTable.Add(0x00, new Callback.Callback(0x00, SetVideoMode));
+        _dispatchTable.Add(0x01, new Callback.Callback(0x01, SetCursorType));
+        _dispatchTable.Add(0x02, new Callback.Callback(0x02, SetCursorPosition));
+        _dispatchTable.Add(0x03, new Callback.Callback(0x03, GetCursorPosition));
+        _dispatchTable.Add(0x05, new Callback.Callback(0x05, SelectActiveDisplayPage));
+        _dispatchTable.Add(0x06, new Callback.Callback(0x06, ScrollPageUp));
+        _dispatchTable.Add(0x07, new Callback.Callback(0x07, ScrollPageDown));
+        _dispatchTable.Add(0x08, new Callback.Callback(0x08, ReadCharacterAndAttributeAtCursor));
+        _dispatchTable.Add(0x09, new Callback.Callback(0x09, WriteCharacterAndAttributeAtCursor));
+        _dispatchTable.Add(0x0A, new Callback.Callback(0x0A, WriteCharacterAtCursor));
+        _dispatchTable.Add(0x0B, new Callback.Callback(0x0B, SetColorPaletteOrBackGroundColor));
+        _dispatchTable.Add(0x0E, new Callback.Callback(0x0E, WriteTextInTeletypeMode));
+        _dispatchTable.Add(0x0F, new Callback.Callback(0x0F, GetVideoMode));
+        _dispatchTable.Add(0x10, new Callback.Callback(0x10, GetSetPaletteRegisters));
+        _dispatchTable.Add(0x11, new Callback.Callback(0x11, CharacterGeneratorRoutine));
+        _dispatchTable.Add(0x12, new Callback.Callback(0x12, VideoSubsystemConfiguration));
+        _dispatchTable.Add(0x13, new Callback.Callback(0x13, WriteString));
+        _dispatchTable.Add(0x1A, new Callback.Callback(0x1A, VideoDisplayCombination));
+        _dispatchTable.Add(0x1B, new Callback.Callback(0x1B, GetFunctionalityInfo));
+    }
+
+    private void WriteString() {
+        uint address = new SegmentedAddress(_state.ES, _state.BP).ToPhysical();
+        StringBuilder res = new();
+        for (uint i = address; i < address + 1024; i++) {
```

Comment:
extracting a string from memory is implemented elsewhere:
https://github.com/OpenRakis/Spice86/blob/master/src/Spice86.Core/Emulator/InterruptHandlers/Dos/DosInt21Handler.cs#L473
https://github.com/OpenRakis/Spice86/blob/master/src/Spice86.Core/Emulator/ReverseEngineer/MemoryBasedDataStructureWithBaseAddressProvider.cs#L35
@maximilien-noal factorized this in his EMS PR by introducing MemoryUtils.GetZeroTerminatedString

Reply:
Yes, I know. Neither current location is usable and the refactor is in another branch.
I also saw it in the EMS branch, but that isn't availabel yet. So I implemented it myself.

I've now pushed a commit that incorporates the code from the EMS branch.